core/src/main/java/com/linecorp/armeria/client/retry/RetryRule.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.retry;
+
+import java.util.concurrent.CompletionStage;
+import java.util.function.Predicate;
+
+import javax.annotation.Nullable;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.UnprocessedRequestException;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.HttpStatusClass;
+import com.linecorp.armeria.common.ResponseHeaders;
+
+/**
+ * A rule for {@link RetryStrategy}.
+ */
+@FunctionalInterface
+public interface RetryRule {
+
+    /**
+     * Adds the idempotent HTTP methods for a {@link RetryStrategy} which will retry
+     * if the request HTTP method is idempotent.
+     */
+    static RetryRuleBuilder onIdempotentMethods() {
```

Comment:
> default() that returns idempotent().onServerError().onException().withBackoff(default)

I agree with adding `default()` method.
I'd like to make the default more conservative and safer to avoid unexpected retry and traffic.

How about retrying only with `UnprocessedRequestException`?
```java
static RetryRule ofDefault() {
  return onUnprocessed().thenBackoff(default);
}
```

Reply:
Yeah that seems like a good default too.

Reply:
Let's add `.onServerError()` as well. 😉

Reply:
For me, this one is the most sensible default:
```java
static RetryRule ofDefault() {
  return onUnprocessed().thenBackoff(default).or(RetryRule.onServerError().onUnprocessed().thenBackoff(default));
}
```

Reply:
Had chat internally on the `default()` method, we could not decide which rule is good for default. And there are other issues as well.
1) A `RetryRule` could not be overridden by other rules. We can combine `RetryRule` using OR operation. That means there is no way to override the default behavior, unlike other defaults.
2) Which is a safe default rule for users? It will depend on the situation. It makes sense that they create an empty rule and add safe rules by their requirements.

So how about just adding a `builder()` method with an empty rule?

Reply:
I personally liked having a one-liner like the previous `onServerErrorStatus` for generic clients, maybe `default` is a bad word but a `simple` (for lack of better word) retry rule that works for many cases and avoids having to go through a huge API seems useful.

By the way, are you planning on keeping all these factories?

Reply:
> a simple (for lack of better word) retry rule that works for many cases and avoids having to go through a huge API seems useful.

Oops, I misrepresented the talk. 😅 Our internal users also want to add a rule that is safe and easy to use, but not default. Let me think about what is good name for it.  🤔

> By the way, are you planning on keeping all these factories?

Yes. as you commented, I am changing the return type of the factory methods to `RetryRule`.
For example:
```java
RetryRule rule = RetryRule.onServerErrorStatus();
```

Reply:
Cool - if the factories return rules that's also helpful for finding a simple one :)

Reply:
The sensible default method named `failsafe()` is added. :-)